cmd/geth/chaincmd.go

```diff
@@ -641,9 +615,99 @@ func parseDumpConfig(ctx *cli.Context, stack *node.Node) (*state.DumpConfig, eth
 		Start:             start.Bytes(),
 		Max:               ctx.Uint64(utils.DumpLimitFlag.Name),
 	}
+
+	db := utils.MakeChainDatabase(ctx, stack, true, false)
+	provided, err := utils.CompareStateSchemeCLIWithConfig(ctx)
+	if err != nil {
+		return nil, nil, common.Hash{}, err
+	}
+	scheme, err := rawdb.ParseStateScheme(provided, db)
+	if err != nil {
+		return nil, nil, common.Hash{}, err
+	}
+	var header *types.Header
+	if scheme == rawdb.PathScheme {
+		triedb := trie.NewDatabase(db, &trie.Config{PathDB: pathdb.ReadOnly})
+		if ctx.NArg() == 1 {
+			arg := ctx.Args().First()
+			if hashish(arg) {
+				hash := common.HexToHash(arg)
+				if number := rawdb.ReadHeaderNumber(db, hash); number != nil {
+					header = rawdb.ReadHeader(db, hash, *number)
+					if header == nil {
+						return nil, nil, common.Hash{}, fmt.Errorf("no head block found")
+					}
+					if contain := triedb.ContainRootHash(header.Root); !contain {
+						return nil, nil, common.Hash{}, fmt.Errorf("PBSS doesn't contain specified MPT root hash for block hash %x", hash)
+					}
+				} else {
+					return nil, nil, common.Hash{}, fmt.Errorf("block hash %x not found", hash)
+				}
+			} else {
+				number, err := strconv.ParseUint(arg, 10, 64)
+				if err != nil {
+					return nil, nil, common.Hash{}, err
+				}
+				if hash := rawdb.ReadCanonicalHash(db, number); hash != (common.Hash{}) {
+					header = rawdb.ReadHeader(db, hash, number)
+					if header == nil {
+						return nil, nil, common.Hash{}, fmt.Errorf("no head block found")
+					}
+					if contain := triedb.ContainRootHash(header.Root); !contain {
```

Comment:
I don't think its necessary to check the roothash here. It adds an extra and useless interface(ContainRootHash) in triedb and brings more maintenance costs for an off-line and low frequency tools.

Reply:
Yes, I'm thinking about this.

Reply:
Refactor this function.